image-rs/src/signature/mechanism/simple/xrss.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2023 IBM Corp.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::*;
+use base64::Engine;
+use oci_distribution::{secrets::RegistryAuth, Reference};
+use reqwest::Client;
+use serde::*;
+
+use crate::signature::{
+    image::{digest::Digest, Image},
+    mechanism::simple,
+};
+
+#[derive(Debug, Default)]
+pub struct RegistryClient {
+    // reqwest client for the container registry that supports the X-R-S-S extension
+    pub client: Client,
+}
+
+#[derive(Deserialize, Debug, PartialEq, Eq, Serialize, Default)]
+pub struct GetOAuthTokenResponse {
+    #[serde(default)]
+    pub token: String,
+    #[serde(default)]
+    pub access_token: String,
+    #[serde(default)]
+    pub expires_in: u32,
+    #[serde(default)]
+    pub issued_at: String,
+}
+
+#[derive(Deserialize, Debug, PartialEq, Eq, Serialize, Default)]
+pub struct GetSignaturesResponse {
+    #[serde(default)]
+    signatures: Vec<RegistrySignature>,
+}
+
+#[derive(Deserialize, Debug, PartialEq, Eq, Serialize, Default)]
+pub struct RegistrySignature {
+    #[serde(default, rename = "schemaVersion")]
+    pub schema_version: u32,
+    #[serde(default)]
+    pub name: String,
+    #[serde(default, rename = "type")]
+    pub signature_type: String,
+    #[serde(default)]
+    pub content: String,
+}
+
+impl RegistryClient {
+    pub fn new() -> RegistryClient {
+        RegistryClient {
+            client: reqwest::Client::new(),
+        }
+    }
+
+    async fn get_oauth_token_for_registry(
+        &self,
+        image: &Image,
+        auth: &RegistryAuth,
+    ) -> Result<String> {
+        match auth {
+            RegistryAuth::Anonymous => {
+                bail!("Trying to get signature from a registry without providing auth information")
+            }
+            RegistryAuth::Basic(username, password) => {
+                let scope = format!("repository:{}:pull", image.reference.repository());
+                let res = self
+                    .client
+                    .get(format!(
+                        "https://{}/oauth/token",
+                        image.reference.registry()
+                    ))
+                    .query(&[
+                        ("account", username),
+                        ("scope", &scope),
+                        ("service", &"registry".to_string()),
+                    ])
+                    .basic_auth(username, Some(password))
+                    .send()
+                    .await
+                    .map_err(|e| anyhow!("Failed to fetch oauth token for registry: {:?}", e))?;
```

Comment:
`anyhow` crate provides a simpler way to add error information context
```suggestion
                    .context("Failed to fetch oauth token for registry")?;
```

Reply:
Changed `.map_err` to `.context`.